tests_e2e/orchestrator/lib/agent_test_suite.py

```diff
@@ -335,24 +312,70 @@ def _setup_node(self, install_test_agent: bool) -> None:
+            try:
+                #
+                # Create a tarball with the files we need to copy to the test node. The tarball includes two directories:
+                #
+                #     * bin - Executables file (Bash and Python scripts)
+                #     * lib - Library files (Python modules)
+                #
+                # After extracting the tarball on the test node, 'bin' will be added to PATH and PYTHONPATH will be set to 'lib'.
+                #
+                # Note that executables are placed directly under 'bin', while the path for Python modules is preserved under 'lib.
+                #
+                tarball_path: Path = Path("/tmp/waagent.tar")
+                log.info("Creating %s with the files need on the test node", tarball_path)
+                log.info("Adding orchestrator/scripts")
+                command = "cd {0} ; tar cvf {1} --transform='s,^,bin/,' *".format(self.context.test_source_directory/"orchestrator"/"scripts", str(tarball_path))
+                log.info("%s\n%s", command, run_command(command, shell=True))
+                log.info("Adding tests/scripts")
+                command = "cd {0} ; tar rvf {1} --transform='s,^,bin/,' *".format(self.context.test_source_directory/"tests"/"scripts", str(tarball_path))
+                log.info("%s\n%s", command, run_command(command, shell=True))
+                log.info("Adding tests/lib")
+                command = "cd {0} ; tar rvf {1} --transform='s,^,lib/,' --exclude=__pycache__ tests_e2e/tests/lib".format(self.context.test_source_directory.parent, str(tarball_path))
+                log.info("%s\n%s", command, run_command(command, shell=True))
+                log.info("Contents of %s:\n\n%s", tarball_path, run_command(['tar', 'tvf', str(tarball_path)]))
+
+                #
+                # Copy the tarball to the test node
+                #
+                log.info("Copying %s to %s:%s", tarball_path, self.context.node.name, target_path)
+                self.context.ssh_client.copy_to_node(tarball_path, target_path)
+
+                #
+                # Extract the tarball and execute the install scripts
+                #
+                log.info('Installing tools on the test node')
+                command = f"tar xvf {target_path/tarball_path.name} && ~/bin/install-tools"
+                log.info("Remote command [%s] completed:\n%s", command, self.context.ssh_client.run_command(command))
+
+                # Tarball creation and extraction was successful - no need to retry
+                tar_retries = 0
+
+            except CommandError as error:
+                if "tar: Unexpected EOF in archive" in error.stderr:
+                    tar_retries -= 1
+                    # Log the error with traceback to see which tar operation failed
+                    log.info(f"Tarball creation or extraction failed: \n{error}")
+                    # Retry tar operations
+                    if tar_retries > 0:
```

Comment:
when we do retry, it's better to cleanup partial tar file if already created during previous attempts

Reply:
just as a comment, the first tar does "cvf".  "c" - creates the file (overwrites any existing file)

Reply:
tar cvf and scp both overwrite any existing files

Reply:
good to know